// @serenity-js/webdriverio/lib/screenplay/models/WebdriverIOPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebdriverIOPage = void 0;
require("webdriverio");
const core_1 = require("@serenity-js/core");
const web_1 = require("@serenity-js/web");
const url_1 = require("url");
const index_js_1 = require("./locators/index.js");
const WebdriverIOCookie_js_1 = require("./WebdriverIOCookie.js");
const WebdriverIOPageElement_js_1 = require("./WebdriverIOPageElement.js");
/**
 * WebdriverIO-specific implementation of {@apilink Page}.
 *
 * @group Models
 */
class WebdriverIOPage extends web_1.Page {
    constructor(session, browser, modalDialogHandler, errorHandler, pageId) {
        super(session, new index_js_1.WebdriverIORootLocator(browser), modalDialogHandler, pageId);
        this.browser = browser;
        this.errorHandler = errorHandler;
    }
    createPageElement(nativeElement) {
        return new WebdriverIOPageElement_js_1.WebdriverIOPageElement(new index_js_1.WebdriverIOExistingElementLocator(this.rootLocator, new web_1.ByCss(String(nativeElement.selector)), this.errorHandler, nativeElement));
    }
    locate(selector) {
        return new WebdriverIOPageElement_js_1.WebdriverIOPageElement(new index_js_1.WebdriverIOLocator(this.rootLocator, selector, this.errorHandler));
    }
    locateAll(selector) {
        return core_1.List.of(new web_1.PageElementsLocator(new index_js_1.WebdriverIOLocator(this.rootLocator, selector, this.errorHandler)));
    }
    async navigateTo(destination) {
        await this.inContextOfThisPage(() => this.browser.url(destination));
    }
    async navigateBack() {
        await this.inContextOfThisPage(() => this.browser.back());
    }
    async navigateForward() {
        await this.inContextOfThisPage(() => this.browser.forward());
    }
    async reload() {
        await this.inContextOfThisPage(() => this.browser.refresh());
    }
    async sendKeys(keys) {
        const keySequence = keys.map(key => {
            if (!web_1.Key.isKey(key)) {
                return key;
            }
            return key.utf16codePoint;
        });
        await this.inContextOfThisPage(() => this.browser.keys(keySequence));
    }
    async executeScript(script, ...args) {
        const innerArguments = [];
        for (const arg of args) {
            const innerArgument = arg instanceof WebdriverIOPageElement_js_1.WebdriverIOPageElement
                ? await arg.nativeElement()
                : arg;
            innerArguments.push(innerArgument);
        }
        const result = await this.inContextOfThisPage(() => {
            return this.browser.execute(script, ...innerArguments);
        });
        this.lastScriptExecutionSummary = new LastScriptExecutionSummary(result);
        return result;
    }
    async executeAsyncScript(script, ...args) {
        const parameters = [];
        for (const arg of args) {
            const parameter = arg instanceof WebdriverIOPageElement_js_1.WebdriverIOPageElement
                ? await arg.nativeElement()
                : arg;
            parameters.push(parameter);
        }
        const result = await this.inContextOfThisPage(() => {
            return this.browser.executeAsync(script, ...parameters);
        });
        this.lastScriptExecutionSummary = new LastScriptExecutionSummary(result);
        return result;
    }
    lastScriptExecutionResult() {
        if (!this.lastScriptExecutionSummary) {
            throw new core_1.LogicError(`Make sure to execute a script before checking on the result`);
        }
        // Selenium returns `null` when the script it executed returns `undefined`
        // so we're mapping the result back.
        return this.lastScriptExecutionSummary.result === null
            ? undefined
            : this.lastScriptExecutionSummary.result;
    }
    async takeScreenshot() {
        return await this.inContextOfThisPage(async () => {
            try {
                return await this.browser.takeScreenshot();
            }
            catch (error) {
                if (error.name === 'ProtocolError' && error.message.includes('Target closed')) {
                    throw new web_1.BrowserWindowClosedError(`Couldn't take screenshot since the browser window is already closed`, error);
                }
                throw error;
            }
        });
    }
    async cookie(name) {
        return new WebdriverIOCookie_js_1.WebdriverIOCookie(this.browser, name);
    }
    async setCookie(cookieData) {
        return await this.inContextOfThisPage(() => {
            return this.browser.setCookies({
                name: cookieData.name,
                value: cookieData.value,
                path: cookieData.path,
                domain: cookieData.domain,
                secure: cookieData.secure,
                httpOnly: cookieData.httpOnly,
                expiry: cookieData.expiry
                    ? cookieData.expiry.toSeconds()
                    : undefined,
                sameSite: cookieData.sameSite,
            });
        });
    }
    async deleteAllCookies() {
        return await this.inContextOfThisPage(() => {
            return this.browser.deleteCookies();
        });
    }
    async title() {
        return await this.inContextOfThisPage(() => this.browser.getTitle());
    }
    async name() {
        return await this.inContextOfThisPage(() => this.browser.execute(`return window.name`));
    }
    async url() {
        return await this.inContextOfThisPage(async () => {
            return new url_1.URL(await this.browser.getUrl());
        });
    }
    async viewportSize() {
        return await this.inContextOfThisPage(async () => {
            if (!this.browser.isDevTools) {
                const calculatedViewportSize = await this.browser.execute(`
                    return {
                        width:  Math.max(document.documentElement.clientWidth,  window.innerWidth || 0),
                        height: Math.max(document.documentElement.clientHeight, window.innerHeight || 0),
                    }
                `);
                // Chrome headless hard-codes window.innerWidth and window.innerHeight to 0
                if (calculatedViewportSize.width > 0 && calculatedViewportSize.height > 0) {
                    return calculatedViewportSize;
                }
            }
            return this.browser.getWindowSize();
        });
    }
    async setViewportSize(size) {
        return await this.inContextOfThisPage(async () => {
            let desiredWindowSize = size;
            if (!this.browser.isDevTools) {
                desiredWindowSize = await this.browser.execute(`
                    var currentViewportWidth  = Math.max(document.documentElement.clientWidth, window.innerWidth || 0)
                    var currentViewportHeight = Math.max(document.documentElement.clientHeight, window.innerHeight || 0)
                    
                    return {
                        width:  Math.max(window.outerWidth  - currentViewportWidth  + ${size.width},  ${size.width}),
                        height: Math.max(window.outerHeight - currentViewportHeight + ${size.height}, ${size.height}),
                    };
                `);
            }
            return this.browser.setWindowSize(desiredWindowSize.width, desiredWindowSize.height);
        });
    }
    async close() {
        await this.inContextOfThisPage(() => this.browser.closeWindow());
    }
    async closeOthers() {
        await this.session.closePagesOtherThan(this);
    }
    async isPresent() {
        const allPages = await this.session.allPages();
        for (const page of allPages) {
            if (page === this) {
                return true;
            }
        }
        return false;
    }
    async inContextOfThisPage(action) {
        let originalCurrentPage;
        try {
            originalCurrentPage = await this.session.currentPage();
            await this.session.changeCurrentPageTo(this);
            return await action();
        }
        catch (error) {
            return await this.errorHandler.executeIfHandled(error, action);
        }
        finally {
            await this.session.changeCurrentPageTo(originalCurrentPage);
        }
    }
}
exports.WebdriverIOPage = WebdriverIOPage;
/**
 * @package
 */
class LastScriptExecutionSummary {
    constructor(result) {
        this.result = result;
    }
}
