// @serenity-js/playwright-test/lib/reporter/SerenityReporterForPlaywrightTest.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.SerenityReporterForPlaywrightTest = void 0;
const core_1 = require("@serenity-js/core");
const events = __importStar(require("@serenity-js/core/lib/events"));
const events_1 = require("@serenity-js/core/lib/events");
const io_1 = require("@serenity-js/core/lib/io");
const model_1 = require("@serenity-js/core/lib/model");
const PlaywrightAttachments_1 = require("./PlaywrightAttachments");
/**
 * Serenity/JS reporter that receives notifications from Playwright Test and emits them as
 * Serenity/JS {@apilink DomainEvent|domain events} which can be used by
 * Serenity/JS {@apilink StageCrewMember|stage crew members}.
 */
class SerenityReporterForPlaywrightTest {
    /**
     * @param config
     * @param serenity
     *  Instance of {@apilink Serenity}, specific to the Node process running this Serenity reporter.
     *  Note that Playwright runs test workers and reporters in separate processes.
     */
    constructor(config, serenity = core_1.serenity) {
        this.serenity = serenity;
        this.errorParser = new PlaywrightErrorParser();
        this.sceneIds = new Map();
        this.serenity.configure(config);
    }
    onBegin(config, suite) {
        this.serenity.announce(new events_1.TestRunStarts(this.now()));
    }
    onTestBegin(test) {
        const currentSceneId = this.serenity.assignNewSceneId();
        this.sceneIds.set(test.id, currentSceneId);
        const scenario = this.scenarioDetailsFrom(test);
        this.emit(new events_1.SceneStarts(currentSceneId, scenario, this.serenity.currentTime()), new events_1.SceneTagged(currentSceneId, new model_1.FeatureTag(scenario.category.value), this.serenity.currentTime()), new events_1.TestRunnerDetected(currentSceneId, new model_1.Name('Playwright'), this.serenity.currentTime()));
    }
    // TODO might be nice to support that by emitting TestStepStarted / Finished
    // onStepBegin(test: TestCase, _result: TestResult, step: TestStep): void {
    //     // console.log('>> onStepBegin');
    // }
    // onStepEnd(test: TestCase, _result: TestResult, step: TestStep): void {
    //     // console.log('>> onStepEnd');
    // }
    onTestEnd(test, result) {
        this.announceRetryIfNeeded(test, result);
        const currentSceneId = this.sceneIds.get(test.id);
        let worstInteractionOutcome = new model_1.ExecutionSuccessful();
        for (const attachment of result.attachments) {
            if (!(attachment.contentType === PlaywrightAttachments_1.SERENITY_JS_DOMAIN_EVENTS_ATTACHMENT_CONTENT_TYPE && attachment.body)) {
                continue;
            }
            const messages = JSON.parse(attachment.body.toString());
            for (const message of messages) {
                if (message.value.sceneId === 'unknown') {
                    message.value.sceneId = currentSceneId.value;
                }
                const event = events[message.type].fromJSON(message.value);
                this.serenity.announce(event);
                if (event instanceof events_1.InteractionFinished && event.outcome.isWorseThan(worstInteractionOutcome)) {
                    worstInteractionOutcome = event.outcome;
                }
            }
        }
        const scenarioOutcome = this.outcomeFrom(test, result);
        this.serenity.announce(new events_1.SceneFinished(currentSceneId, this.scenarioDetailsFrom(test), this.determineScenarioOutcome(worstInteractionOutcome, scenarioOutcome), this.now()));
    }
    determineScenarioOutcome(worstInteractionOutcome, scenarioOutcome) {
        if (worstInteractionOutcome instanceof model_1.ExecutionFailedWithAssertionError) {
            return worstInteractionOutcome;
        }
        return worstInteractionOutcome.isWorseThan(scenarioOutcome)
            ? worstInteractionOutcome
            : scenarioOutcome;
    }
    outcomeFrom(test, result) {
        const outcome = test.outcome();
        if (outcome === 'skipped') {
            return new model_1.ExecutionSkipped();
        }
        if (outcome === 'unexpected' && result.status === 'passed') {
            return new model_1.ExecutionFailedWithError(new core_1.LogicError(`Scenario expected to fail, but ${result.status}`));
        }
        if (['failed', 'interrupted', 'timedOut'].includes(result.status)) {
            if (test.retries > result.retry) {
                return new model_1.ExecutionIgnored(this.errorParser.errorFrom(result.error));
            }
            return new model_1.ExecutionFailedWithError(this.errorParser.errorFrom(result.error));
        }
        return new model_1.ExecutionSuccessful();
    }
    scenarioDetailsFrom(test) {
        const [root_, browserName_, fileName, describeOrItBlockTitle, ...nestedTitles] = test.titlePath();
        const path = new io_1.Path(test.location.file);
        const scenarioName = nestedTitles.join(' ').trim();
        const featureName = scenarioName
            ? describeOrItBlockTitle
            : fileName;
        return new model_1.ScenarioDetails(new model_1.Name(scenarioName || describeOrItBlockTitle), new model_1.Category(featureName), new io_1.FileSystemLocation(path, test.location.line, test.location.column));
    }
    async onEnd() {
        this.serenity.announce(new events_1.TestRunFinishes(this.serenity.currentTime()));
        try {
            await this.serenity.waitForNextCue();
            this.serenity.announce(new events_1.TestRunFinished(new model_1.ExecutionSuccessful(), this.serenity.currentTime()));
        }
        catch (error) {
            this.serenity.announce(new events_1.TestRunFinished(new model_1.ExecutionFailedWithError(error), this.serenity.currentTime()));
            throw error;
        }
    }
    // TODO emit a text artifact with stdout?
    // reporter.onStdErr(chunk, test, result)
    // reporter.onStdOut(chunk, test, result)
    emit(...events) {
        events.forEach(event => {
            this.serenity.announce(event);
        });
    }
    announceRetryIfNeeded(test, result) {
        if (test.retries === 0) {
            return;
        }
        const currentSceneId = this.sceneIds.get(test.id);
        this.emit(new events_1.RetryableSceneDetected(currentSceneId, this.now()), new events_1.SceneTagged(currentSceneId, new model_1.ArbitraryTag('retried'), // todo: replace with a dedicated tag
        this.now()));
        if (result.retry > 0) {
            this.emit(new events_1.SceneTagged(currentSceneId, new model_1.ExecutionRetriedTag(result.retry), this.serenity.currentTime()));
        }
    }
    now() {
        return this.serenity.currentTime();
    }
    printsToStdio() {
        return true;
    }
}
exports.SerenityReporterForPlaywrightTest = SerenityReporterForPlaywrightTest;
class PlaywrightErrorParser {
    errorFrom(testError) {
        const message = testError.message && PlaywrightErrorParser.stripAsciiFrom(testError.message);
        let stack = testError.stack && PlaywrightErrorParser.stripAsciiFrom(testError.stack);
        // TODO: Do I need to process it?
        // const value     = testError.value;
        const prologue = `Error: ${message}`;
        if (stack && message && stack.startsWith(prologue)) {
            stack = stack.slice(prologue.length);
        }
        const error = new Error(message);
        error.stack = stack;
        return error;
    }
    static stripAsciiFrom(text) {
        return text.replace(this.ascii, '');
    }
}
PlaywrightErrorParser.ascii = new RegExp('[\\u001B\\u009B][[\\]()#;?]*(?:(?:(?:[a-zA-Z\\d]*(?:;[-a-zA-Z\\d\\/#&.:=?%@~_]*)*)?\\u0007)|(?:(?:\\d{1,4}(?:;\\d{0,4})*)?[\\dA-PR-TZcf-ntqry=><~]))', // eslint-disable-line no-control-regex
'g');
