// @serenity-js/playwright/lib/screenplay/models/PlaywrightBrowsingSessionWithPage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaywrightBrowsingSessionWithPage = void 0;
const model_1 = require("@serenity-js/core/lib/model");
const PlaywrightBrowsingSession_1 = require("./PlaywrightBrowsingSession");
const PlaywrightPage_1 = require("./PlaywrightPage");
/**
 *  @group Models
 */
class PlaywrightBrowsingSessionWithPage extends PlaywrightBrowsingSession_1.PlaywrightBrowsingSession {
    constructor(page, browserContextOptions) {
        super(browserContextOptions);
        this.page = page;
        this.currentBrowserPage = new PlaywrightPage_1.PlaywrightPage(this, this.page, this.browserContextOptions, model_1.CorrelationId.create());
    }
    async createBrowserContext(options) {
        return this.page.context();
    }
    async browserCapabilities() {
        const browser = await this.page.context().browser();
        return {
            browserName: browser._initializer.name,
            platformName: process.platform,
            browserVersion: browser.version()
        };
    }
}
exports.PlaywrightBrowsingSessionWithPage = PlaywrightBrowsingSessionWithPage;
