// @serenity-js/playwright/lib/screenplay/models/PlaywrightBrowsingSessionWithBrowser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PlaywrightBrowsingSessionWithBrowser = void 0;
const selector_engines_1 = require("../../selector-engines");
const PlaywrightBrowsingSession_1 = require("./PlaywrightBrowsingSession");
/**
 *  @group Models
 */
class PlaywrightBrowsingSessionWithBrowser extends PlaywrightBrowsingSession_1.PlaywrightBrowsingSession {
    constructor(browser, browserContextOptions, selectors) {
        super(browserContextOptions);
        this.browser = browser;
        this.selectors = selectors;
        this.serenitySelectorEngines = new selector_engines_1.SerenitySelectorEngines();
    }
    async createBrowserContext(options) {
        await this.serenitySelectorEngines.ensureRegisteredWith(this.selectors);
        return this.browser.newContext(this.browserContextOptions);
    }
    async closeAllPages() {
        await super.closeAllPages();
        const context = await this.browserContext();
        await context.close();
    }
    async browserCapabilities() {
        return {
            browserName: this.browser._initializer.name,
            platformName: process.platform,
            browserVersion: this.browser.version()
        };
    }
}
exports.PlaywrightBrowsingSessionWithBrowser = PlaywrightBrowsingSessionWithBrowser;
