// @serenity-js/serenity-bdd/lib/stage/crew/serenity-bdd-reporter/processors/transformations/scenarioDetailsOf.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.scenarioDetailsOf = void 0;
const tiny_types_1 = require("tiny-types");
const mappers_1 = require("../mappers");
/**
 * @package
 */
function scenarioDetailsOf(details) {
    return (context) => {
        const name = (0, tiny_types_1.ensure)('scenario name', details.name.value, (0, tiny_types_1.isNotBlank)());
        const category = (0, tiny_types_1.ensure)('scenario category', details.category.value, (0, tiny_types_1.isNotBlank)());
        context.report.name = (0, mappers_1.escapeHtml)(name);
        context.report.title = (0, mappers_1.escapeHtml)(name);
        context.report.manual = false;
        context.report.testSteps = [];
        context.report.userStory = {
            id: (0, mappers_1.dashify)(category),
            storyName: (0, mappers_1.escapeHtml)(category),
            path: isFeatureFile(details.location.path)
                ? details.location.path.value
                : '',
            type: 'feature',
        };
        return context;
    };
}
exports.scenarioDetailsOf = scenarioDetailsOf;
/**
 * @package
 * @param {Path} path
 */
function isFeatureFile(path) {
    return path && path.value.endsWith('.feature');
}
